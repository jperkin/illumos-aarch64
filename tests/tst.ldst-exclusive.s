/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2017 Joyent, Inc.
 */

/*
 * C3.3.6 Load/store exclusive
 *
 * Test the various combinations of register arguments:
 *
 *   OP <32> <32>            [..
 *   OP      <32>            [..
 *   OP <32> <32|64>         [..
 *   OP <32> <32|64> <32|64> [..
 *   OP      <32|64>         [..
 *   OP      <32|64> <32|64> [..
 *
 * #0 is optional and default.
 */

.macro ldstexcl1 op test64=0
	\op	w0,  [x0]
	\op	w30, [x0, #0]
	\op	wzr, [sp]
	\op	wzr, [sp, #0]
	.if \test64
	\op	x0,  [x0]
	\op	x30, [x0, #0]
	\op	xzr, [sp]
	\op	xzr, [sp, #0]
	.endif
.endm

.macro ldstexcl2 op test64_1=0 test64_2=0
	\op	w0,  w0,  [x0]
	\op	w30, w30, [x0, #0]
	\op	wzr, wzr, [sp]
	\op	wzr, wzr, [sp, #0]
	.if \test64_1
	\op	w0,  x0,  [x0]
	\op	w30, x30, [x0, #0]
	\op	wzr, xzr, [sp]
	\op	wzr, xzr, [sp, #0]
	.endif
	.if \test64_2
	\op	x0,  x0,  [x0]
	\op	x30, x30, [x0, #0]
	\op	xzr, xzr, [sp]
	\op	xzr, xzr, [sp, #0]
	.endif
.endm

.macro ldstexcl3 op
	\op	w0,  w15, w30, [x0]
	\op	w30, w15, w0,  [x0, #0]
	\op	wzr, wzr, wzr, [sp]
	\op	wzr, wzr, wzr, [sp, #0]
	\op	w0,  x15, x30, [x0]
	\op	w30, x15, x0,  [x0, #0]
	\op	wzr, xzr, xzr, [sp]
	\op	wzr, xzr, xzr, [sp, #0]
.endm

.text
.globl libdis_test
.type libdis_test, @function
libdis_test:
	ldstexcl2 stxrb
	ldstexcl2 stlxrb
	ldstexcl1 ldxrb
	ldstexcl1 ldaxrb
	ldstexcl1 stlrb
	ldstexcl1 ldarb
	ldstexcl2 stxrh
	ldstexcl2 stlxrh
	ldstexcl1 ldxrh
	ldstexcl1 ldaxrh
	ldstexcl1 stlrh
	ldstexcl1 ldarh
	ldstexcl2 stxr 1
	ldstexcl2 stlxr 1
	ldstexcl3 stxp
	ldstexcl3 stlxp
	ldstexcl1 ldxr 1
	ldstexcl1 ldaxr 1
	ldstexcl2 ldxp 0 1
	ldstexcl2 ldaxp 0 1
	ldstexcl1 stlr 1
	ldstexcl1 ldar 1
.size libdis_test, [.-libdis_test]
