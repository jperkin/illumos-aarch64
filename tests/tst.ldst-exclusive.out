    libdis_test:       00 7c 00 08  stxrb w0, w0, [x0]
    libdis_test+0x4:   1e 7c 1e 08  stxrb w30, w30, [x0]
    libdis_test+0x8:   ff 7f 1f 08  stxrb wzr, wzr, [sp]
    libdis_test+0xc:   ff 7f 1f 08  stxrb wzr, wzr, [sp]
    libdis_test+0x10:  00 fc 00 08  stlxrb w0, w0, [x0]
    libdis_test+0x14:  1e fc 1e 08  stlxrb w30, w30, [x0]
    libdis_test+0x18:  ff ff 1f 08  stlxrb wzr, wzr, [sp]
    libdis_test+0x1c:  ff ff 1f 08  stlxrb wzr, wzr, [sp]
    libdis_test+0x20:  00 7c 5f 08  ldxrb w0, [x0]
    libdis_test+0x24:  1e 7c 5f 08  ldxrb w30, [x0]
    libdis_test+0x28:  ff 7f 5f 08  ldxrb wzr, [sp]
    libdis_test+0x2c:  ff 7f 5f 08  ldxrb wzr, [sp]
    libdis_test+0x30:  00 fc 5f 08  ldaxrb w0, [x0]
    libdis_test+0x34:  1e fc 5f 08  ldaxrb w30, [x0]
    libdis_test+0x38:  ff ff 5f 08  ldaxrb wzr, [sp]
    libdis_test+0x3c:  ff ff 5f 08  ldaxrb wzr, [sp]
    libdis_test+0x40:  00 fc 9f 08  stlrb w0, [x0]
    libdis_test+0x44:  1e fc 9f 08  stlrb w30, [x0]
    libdis_test+0x48:  ff ff 9f 08  stlrb wzr, [sp]
    libdis_test+0x4c:  ff ff 9f 08  stlrb wzr, [sp]
    libdis_test+0x50:  00 fc df 08  ldarb w0, [x0]
    libdis_test+0x54:  1e fc df 08  ldarb w30, [x0]
    libdis_test+0x58:  ff ff df 08  ldarb wzr, [sp]
    libdis_test+0x5c:  ff ff df 08  ldarb wzr, [sp]
    libdis_test+0x60:  00 7c 00 48  stxrh w0, w0, [x0]
    libdis_test+0x64:  1e 7c 1e 48  stxrh w30, w30, [x0]
    libdis_test+0x68:  ff 7f 1f 48  stxrh wzr, wzr, [sp]
    libdis_test+0x6c:  ff 7f 1f 48  stxrh wzr, wzr, [sp]
    libdis_test+0x70:  00 fc 00 48  stlxrh w0, w0, [x0]
    libdis_test+0x74:  1e fc 1e 48  stlxrh w30, w30, [x0]
    libdis_test+0x78:  ff ff 1f 48  stlxrh wzr, wzr, [sp]
    libdis_test+0x7c:  ff ff 1f 48  stlxrh wzr, wzr, [sp]
    libdis_test+0x80:  00 7c 5f 48  ldxrh w0, [x0]
    libdis_test+0x84:  1e 7c 5f 48  ldxrh w30, [x0]
    libdis_test+0x88:  ff 7f 5f 48  ldxrh wzr, [sp]
    libdis_test+0x8c:  ff 7f 5f 48  ldxrh wzr, [sp]
    libdis_test+0x90:  00 fc 5f 48  ldaxrh w0, [x0]
    libdis_test+0x94:  1e fc 5f 48  ldaxrh w30, [x0]
    libdis_test+0x98:  ff ff 5f 48  ldaxrh wzr, [sp]
    libdis_test+0x9c:  ff ff 5f 48  ldaxrh wzr, [sp]
    libdis_test+0xa0:  00 fc 9f 48  stlrh w0, [x0]
    libdis_test+0xa4:  1e fc 9f 48  stlrh w30, [x0]
    libdis_test+0xa8:  ff ff 9f 48  stlrh wzr, [sp]
    libdis_test+0xac:  ff ff 9f 48  stlrh wzr, [sp]
    libdis_test+0xb0:  00 fc df 48  ldarh w0, [x0]
    libdis_test+0xb4:  1e fc df 48  ldarh w30, [x0]
    libdis_test+0xb8:  ff ff df 48  ldarh wzr, [sp]
    libdis_test+0xbc:  ff ff df 48  ldarh wzr, [sp]
    libdis_test+0xc0:  00 7c 00 88  stxr w0, w0, [x0]
    libdis_test+0xc4:  1e 7c 1e 88  stxr w30, w30, [x0]
    libdis_test+0xc8:  ff 7f 1f 88  stxr wzr, wzr, [sp]
    libdis_test+0xcc:  ff 7f 1f 88  stxr wzr, wzr, [sp]
    libdis_test+0xd0:  00 7c 00 c8  stxr w0, x0, [x0]
    libdis_test+0xd4:  1e 7c 1e c8  stxr w30, x30, [x0]
    libdis_test+0xd8:  ff 7f 1f c8  stxr wzr, xzr, [sp]
    libdis_test+0xdc:  ff 7f 1f c8  stxr wzr, xzr, [sp]
    libdis_test+0xe0:  00 fc 00 88  stlxr w0, w0, [x0]
    libdis_test+0xe4:  1e fc 1e 88  stlxr w30, w30, [x0]
    libdis_test+0xe8:  ff ff 1f 88  stlxr wzr, wzr, [sp]
    libdis_test+0xec:  ff ff 1f 88  stlxr wzr, wzr, [sp]
    libdis_test+0xf0:  00 fc 00 c8  stlxr w0, x0, [x0]
    libdis_test+0xf4:  1e fc 1e c8  stlxr w30, x30, [x0]
    libdis_test+0xf8:  ff ff 1f c8  stlxr wzr, xzr, [sp]
    libdis_test+0xfc:  ff ff 1f c8  stlxr wzr, xzr, [sp]
    libdis_test+0x100: 0f 78 20 88  stxp w0, w15, w30, [x0]
    libdis_test+0x104: 0f 00 3e 88  stxp w30, w15, w0, [x0]
    libdis_test+0x108: ff 7f 3f 88  stxp wzr, wzr, wzr, [sp]
    libdis_test+0x10c: ff 7f 3f 88  stxp wzr, wzr, wzr, [sp]
    libdis_test+0x110: 0f 78 20 c8  stxp w0, x15, x30, [x0]
    libdis_test+0x114: 0f 00 3e c8  stxp w30, x15, x0, [x0]
    libdis_test+0x118: ff 7f 3f c8  stxp wzr, xzr, xzr, [sp]
    libdis_test+0x11c: ff 7f 3f c8  stxp wzr, xzr, xzr, [sp]
    libdis_test+0x120: 0f f8 20 88  stlxp w0, w15, w30, [x0]
    libdis_test+0x124: 0f 80 3e 88  stlxp w30, w15, w0, [x0]
    libdis_test+0x128: ff ff 3f 88  stlxp wzr, wzr, wzr, [sp]
    libdis_test+0x12c: ff ff 3f 88  stlxp wzr, wzr, wzr, [sp]
    libdis_test+0x130: 0f f8 20 c8  stlxp w0, x15, x30, [x0]
    libdis_test+0x134: 0f 80 3e c8  stlxp w30, x15, x0, [x0]
    libdis_test+0x138: ff ff 3f c8  stlxp wzr, xzr, xzr, [sp]
    libdis_test+0x13c: ff ff 3f c8  stlxp wzr, xzr, xzr, [sp]
    libdis_test+0x140: 00 7c 5f 88  ldxr w0, [x0]
    libdis_test+0x144: 1e 7c 5f 88  ldxr w30, [x0]
    libdis_test+0x148: ff 7f 5f 88  ldxr wzr, [sp]
    libdis_test+0x14c: ff 7f 5f 88  ldxr wzr, [sp]
    libdis_test+0x150: 00 7c 5f c8  ldxr x0, [x0]
    libdis_test+0x154: 1e 7c 5f c8  ldxr x30, [x0]
    libdis_test+0x158: ff 7f 5f c8  ldxr xzr, [sp]
    libdis_test+0x15c: ff 7f 5f c8  ldxr xzr, [sp]
    libdis_test+0x160: 00 fc 5f 88  ldaxr w0, [x0]
    libdis_test+0x164: 1e fc 5f 88  ldaxr w30, [x0]
    libdis_test+0x168: ff ff 5f 88  ldaxr wzr, [sp]
    libdis_test+0x16c: ff ff 5f 88  ldaxr wzr, [sp]
    libdis_test+0x170: 00 fc 5f c8  ldaxr x0, [x0]
    libdis_test+0x174: 1e fc 5f c8  ldaxr x30, [x0]
    libdis_test+0x178: ff ff 5f c8  ldaxr xzr, [sp]
    libdis_test+0x17c: ff ff 5f c8  ldaxr xzr, [sp]
    libdis_test+0x180: 00 00 7f 88  ldxp w0, w0, [x0]
    libdis_test+0x184: 1e 78 7f 88  ldxp w30, w30, [x0]
    libdis_test+0x188: ff 7f 7f 88  ldxp wzr, wzr, [sp]
    libdis_test+0x18c: ff 7f 7f 88  ldxp wzr, wzr, [sp]
    libdis_test+0x190: 00 00 7f c8  ldxp x0, x0, [x0]
    libdis_test+0x194: 1e 78 7f c8  ldxp x30, x30, [x0]
    libdis_test+0x198: ff 7f 7f c8  ldxp xzr, xzr, [sp]
    libdis_test+0x19c: ff 7f 7f c8  ldxp xzr, xzr, [sp]
    libdis_test+0x1a0: 00 80 7f 88  ldaxp w0, w0, [x0]
    libdis_test+0x1a4: 1e f8 7f 88  ldaxp w30, w30, [x0]
    libdis_test+0x1a8: ff ff 7f 88  ldaxp wzr, wzr, [sp]
    libdis_test+0x1ac: ff ff 7f 88  ldaxp wzr, wzr, [sp]
    libdis_test+0x1b0: 00 80 7f c8  ldaxp x0, x0, [x0]
    libdis_test+0x1b4: 1e f8 7f c8  ldaxp x30, x30, [x0]
    libdis_test+0x1b8: ff ff 7f c8  ldaxp xzr, xzr, [sp]
    libdis_test+0x1bc: ff ff 7f c8  ldaxp xzr, xzr, [sp]
    libdis_test+0x1c0: 00 fc 9f 88  stlr w0, [x0]
    libdis_test+0x1c4: 1e fc 9f 88  stlr w30, [x0]
    libdis_test+0x1c8: ff ff 9f 88  stlr wzr, [sp]
    libdis_test+0x1cc: ff ff 9f 88  stlr wzr, [sp]
    libdis_test+0x1d0: 00 fc 9f c8  stlr x0, [x0]
    libdis_test+0x1d4: 1e fc 9f c8  stlr x30, [x0]
    libdis_test+0x1d8: ff ff 9f c8  stlr xzr, [sp]
    libdis_test+0x1dc: ff ff 9f c8  stlr xzr, [sp]
    libdis_test+0x1e0: 00 fc df 88  ldar w0, [x0]
    libdis_test+0x1e4: 1e fc df 88  ldar w30, [x0]
    libdis_test+0x1e8: ff ff df 88  ldar wzr, [sp]
    libdis_test+0x1ec: ff ff df 88  ldar wzr, [sp]
    libdis_test+0x1f0: 00 fc df c8  ldar x0, [x0]
    libdis_test+0x1f4: 1e fc df c8  ldar x30, [x0]
    libdis_test+0x1f8: ff ff df c8  ldar xzr, [sp]
    libdis_test+0x1fc: ff ff df c8  ldar xzr, [sp]
